google/cloud/pubsub/internal/tracing_message_batch.cc

```diff
@@ -35,66 +35,86 @@ void TracingMessageBatch::SaveMessage(pubsub::Message m) {
+  auto context = batch_sink_parent_span->GetContext();
+  auto trace_id = internal::ToString(context.trace_id());
+  auto span_id = internal::ToString(context.span_id());
+  for (auto& message_span : message_spans) {
     message_span->AddEvent("gl-cpp.batch_flushed");
+    message_span->SetAttribute("pubsub.batch_sink.trace_id", trace_id);
+    message_span->SetAttribute("pubsub.batch_sink.span_id", span_id);
   }
 }
 
-void TracingMessageBatch::Flush() {
+opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> MakeParentSpan(
+    std::vector<
+        opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>> const&
+        message_spans) {
   using opentelemetry::trace::SpanContext;
   using AttributesList =
       std::vector<std::pair<opentelemetry::nostd::string_view,
                             opentelemetry::common::AttributeValue>>;
   auto constexpr kMaxOtelLinks = 128;
   std::vector<std::pair<SpanContext, AttributesList>> links;
-  auto batch_size = message_spans_.size();
+  auto batch_size = message_spans.size();
   links.reserve(batch_size);
-
   // If the batch size is less than the max size, add the links to a single
   // span.
   if (batch_size < kMaxOtelLinks) {
     std::transform(
-        message_spans_.begin(), message_spans_.end(), std::back_inserter(links),
+        message_spans.begin(), message_spans.end(), std::back_inserter(links),
         [i = static_cast<std::int64_t>(0)](auto const& span) mutable {
           return std::make_pair(
               span->GetContext(),
               AttributesList{{"messaging.pubsub.message.link", i++}});
         });
   }
-  auto batch_sink_span_parent =
+  auto batch_sink_parent_span =
       internal::MakeSpan("BatchSink::AsyncPublish",
                          /*attributes=*/
                          {{"messaging.pubsub.num_messages_in_batch",
                            static_cast<std::int64_t>(batch_size)}},
                          /*links*/ links);
+  return batch_sink_parent_span;
+}
 
-  // TODO(#12528): Handle batches larger than 128.
+void TracingMessageBatch::Flush() {
+  decltype(message_spans_) message_spans;
+  {
+    std::lock_guard<std::mutex> lk(message_mu_);
+    message_spans.swap(message_spans_);
+  }
 
-  // This must be called before we clear the message spans.
-  AddMessageSpanMetadata();
-  // Clear message spans.
-  message_spans_.clear();
+  auto batch_sink_parent_span = MakeParentSpan(message_spans);
 
-  batch_sink_spans_.push_back(batch_sink_span_parent);
+  // TODO(#12528): Handle batches larger than 128.
+  AddMessageSpanMetadata(batch_sink_parent_span, std::move(message_spans));
 
-  // Set the batch sink parent span.
+  // Set the batch sink as the active span.
   auto async_scope = internal::GetTracer(internal::CurrentOptions())
-                         ->WithActiveSpan(batch_sink_span_parent);
+                         ->WithActiveSpan(batch_sink_parent_span);
+  {
+    std::lock_guard<std::mutex> lk(batch_sink_mu_);
```

Comment:
I see what you did with the locking and mutexes. Very nice.

There are cases where fewer locks with longer critical sections is better (e.g. if there is almost no contention for the locks and the number of locks has grown way beyond 2 or 3). It would be shocking if that was the case here.

Reply:
Ack. I'll keep that in mind for the future.